### DISCUSSION TITLE
Learning tools interoperability (LTI) - Support

### DISCUSSION THREAD

**Discussion by NilsBaumgartner1994** [directus/directus#16549] [Feature Requests]
Is it possible to support a Learning tools interoperability (LTI)?

**Comment by NilsBaumgartner1994** [comment]
My idea would be: the lti tool directs to a Directus SSO button: which then redirects to an SSO mock server which takes the informations given from Directus relay with the post informations and checks it against the outcome service. On a successful validation the SSO service then redirects to Directus.

**Comment by rijkvanzanten** [comment]
Mind elaborating on this idea? What is LTI? How would it integrate? What does it do? How do you imagine Directus would utilize this?

**Reply by NilsBaumgartner1994** [reply]
Sure. LTI stands for Learning tools interoperability. I will picture a small use case for the Directus support. 

In our university I have my students as a teacher. We also have a web-platform for groups, courses, mails and so on. We have the ability to add new tabs to a course online adding some predefined tools. One of them allows to embed iFrames and LTI-Tools.

since Directus does not support iFrame out of the box (found a issue somewhere) and in my tests I could not get it to work. But working with a custom Frontend Application using Directus in the Backend works too.

But back to the LTI. Adding an LTI (iFrame) brings the following benefits. The Learn Management Platform which supports LTI will pass Informations to the iFrame.

The LTI Application will informations such as: 
- The username
- The course
- additionAl informations 
- Role as teacher or student
- A route to verify the informations

I think the main concept here is an oAuth1 authentication. Regarding that Directus supports the awesome oAuth2 is the oAuth1 not very cool. 

I think in order to support the LTI Feature the following changes need to be considered:
- Directus_User Auth Provider would be related to the course: „University_Databsecourse_A“
- Adding the URL of the Learning Platform to encounter „Hacking“ (similar to oAuth2 registering allowed services)

Workaround:
- Currently we build a small application which „mocks“ an oAuth2 Server which will register the user with the given and verified Informations passed from the LTI Tool.

Benefits:
- Directus could be established in the University and School context easier, which allows to let students and colleagues build easier applications
- Users don’t need another account to create using an LTI tool

https://www.imsglobal.org/activity/learning-tools-interoperability

**Reply by danielband** [reply]
@NilsBaumgartner1994 Interesting, I work in private education sector, that´s why this post immediately grabbed my attention. How are you using Directus? As a course management platform or as an application backend?  You mentioned something about using iframes in Directus, so my guess is, that you want to use it to manage courses.

Is it possible for you guys to use LTI1.3 which supports oauth2?

**Reply by rijkvanzanten** [reply]
Cool, I've never heard of this! So just so I understand completely; this would effectively be another way to login like OAuth2, but as a more specialized auth flow to make it LTI compatible, or are there other features that you'd hope to see on the Directus side of the equation?

**Reply by NilsBaumgartner1994** [reply]
> @NilsBaumgartner1994 Interesting, I work in private education sector, that´s why this post immediately grabbed my attention. How are you using Directus? As a course management platform or as an application backend? You mentioned something about using iframes in Directus, so my guess is, that you want to use it to manage courses.
> 
> Is it possible for you guys to use LTI1.3 which supports oauth2?

Hi daniel, nice to hear that this topic interest you too. To be specific we use here at our university (and most universitied) Stud.IP. Stud.IP does not provide oAuth2 itself. Unfortunately I think they only support the LTI 1.X (https://hilfe.studip.de/admin/GlobalEinstellungen/LTI-Tools). I dont think that this supports oAuth2.

I tested the LTI with: https://www.tsugi.org/lti-test/tool.php


<img width="703" alt="Bildschirmfoto 2023-01-11 um 00 23 38" src="https://user-images.githubusercontent.com/33295329/211683010-6eb475df-7d21-4c29-a99f-167aeaeac90b.png">

<img width="668" alt="Bildschirmfoto 2023-01-11 um 00 27 15" src="https://user-images.githubusercontent.com/33295329/211683437-55462bd3-9647-46be-a84f-4c131733ed39.png">

**Reply by NilsBaumgartner1994** [reply]
> Cool, I've never heard of this! So just so I understand completely; this would effectively be another way to login like OAuth2, but as a more specialized auth flow to make it LTI compatible, or are there other features that you'd hope to see on the Directus side of the equation?

"n the current version of the specification, v1.3, this is done using [OAuth2](https://en.wikipedia.org/wiki/OAuth2), [OpenID Connect](https://en.wikipedia.org/wiki/OpenID_Connect), and [JSON Web Tokens](https://en.wikipedia.org/wiki/JSON_Web_Token)".
But many many Applications supporting LTI have the older v1.X version which does not supports oAuth2.

So basicly it is either oAuth1 or oAuth2 but nothing very special. I think the auth flow is a bit different, since the application will receive the information where to get the oAuth flow informations. I am not very into this topic.

I attached an Image what informations the LTI application would receive.


![](https://user-images.githubusercontent.com/33295329/211683437-55462bd3-9647-46be-a84f-4c131733ed39.png)

**Reply by paescuj** [reply]
So if I understand correctly, it boils down to add support for OAuth 1 in Directus. I'm not sure that's something we should be pursuing. OAuth 1 is a very old standard which is gradually used less, many providers already have or are soon deprecating it and I'm also afraid we won't find a well-maintained Node.js library which we could rely on.

In principle, however, your use case sounds interesting! Would there possibly be other ways for integration?

**Reply by NilsBaumgartner1994** [reply]
> So if I understand correctly, it boils down to add support for OAuth 1 in Directus. I'm not sure that's something we should be pursuing. OAuth 1 is a very old standard which is gradually used less, many providers already have or are soon deprecating it and I'm also afraid we won't find a well-maintained Node.js library which we could rely on.
> 
> In principle, however, your use case sounds interesting! Would there possibly be other ways for integration?

Solutions would be to force the platform to support LTI 1.3. a workaround would be to mock an oAuth2 server. I also won’t recommend supporting oAuth1 since it is old and there is a oAuth2 solution.

But a general support in Directus for LTI 1.3 would be nice.
The use cases would be very nice :-)

**Reply by rijkvanzanten** [reply]
Does anybody here happen to have a decent reference implementation? I found the full spec, but it's a little tricky for me to guestimate what sort of level of effort we're talking about here just based on that 😅 https://www.imsglobal.org/spec/lti/v1p3

**Reply by NilsBaumgartner1994** [reply]
> Does anybody here happen to have a decent reference implementation? I found the full spec, but it's a little tricky for me to guestimate what sort of level of effort we're talking about here just based on that 😅 https://www.imsglobal.org/spec/lti/v1p3

I could only get an implementation for the old LTI 1.1 Tool. But that does not help very much :D

**Reply by danielband** [reply]
> Does anybody here happen to have a decent reference implementation? I found the full spec, but it's a little tricky for me to guestimate what sort of level of effort we're talking about here just based on that 😅 https://www.imsglobal.org/spec/lti/v1p3

@rijkvanzanten [here](https://github.com/1EdTech/ltibootcamp) you go. LTI is basically a standard about displaying external content from an "LTI-Tool" in a Learning Management System (LMS) like Moodle or Stud.IP. Authentication is just one of the parts that changed in LTI1.3. 

If I understand Nils correctly, Directus would act like a "Tool builder". @NilsBaumgartner1994  is that correct?

So the flow would be something like this:
1. On user interaction, the LMS launches an oauth2 flow that reaches Directus. 
2. Directus would validate the issuer of that request and on success create an auth request, including a redirect url to the tool, and send that back to the LMS. 
3. In turn the LMS would validate the redirect url and send an id token to Directus
4. After checking the validity of the token and checking access control. On success the user would be able to access the external content inside of the LMS.

I don´t understand few things though.. 
1) Why would LMS users need to authenticate with Directus?
2) Where is the business logic of the Tool?
3) Does Directus serve the static assets of the Tool?
4) How will creators develop a Tool?

**Reply by NilsBaumgartner1994** [reply]
> > Does anybody here happen to have a decent reference implementation? I found the full spec, but it's a little tricky for me to guestimate what sort of level of effort we're talking about here just based on that 😅 https://www.imsglobal.org/spec/lti/v1p3
> 
> @rijkvanzanten [here](https://github.com/1EdTech/ltibootcamp) you go. LTI is basically a standard about displaying external content from an "LTI-Tool" in a Learning Management System (LMS) like Moodle or Stud.IP. Authentication is just one of the parts that changed in LTI1.3.
> 
> If I understand Nils correctly, Directus would act like a "Tool builder". @NilsBaumgartner1994 is that correct?
> 
> So the flow would be something like this:
> 
> 1. On user interaction, the LMS launches an oauth2 flow that reaches Directus.
> 2. Directus would validate the issuer of that request and on success create an auth request, including a redirect url to the tool, and send that back to the LMS.
> 3. In turn the LMS would validate the redirect url and send an id token to Directus
> 4. After checking the validity of the token and checking access control. On success the user would be able to access the external content inside of the LMS.
> 
> I don´t understand few things though..
> 
> 1. Why would LMS users need to authenticate with Directus?
> 2. Where is the business logic of the Tool?
> 3. Does Directus serve the static assets of the Tool?
> 4. How will creators develop a Tool?

Correct, Directus would act like a Tool builder. In generally it would just authenticate the users. After that we could use directus inside an LTI-Platform or redirect the user: With a custom redirectHook (#8867) to our own frontend.

**Reply by danielband** [reply]
I am thinking.. Actually Directus does not need to authenticate anything, since the LMS should authenticate the user and pass the info to the Directus instance. So, wouldn´t it be possible to create custom endpoints, bypassing directus auth and issuing a token to the client?

**Reply by NilsBaumgartner1994** [reply]
> I am thinking.. Actually Directus does not need to authenticate anything, since the LMS should authenticate the user and pass the info to the Directus instance. So, wouldn´t it be possible to create custom endpoints, bypassing directus auth and issuing a token to the client?

The authentication would be needed in LTI 1.1. For LTI 1.3 with oAuth2 there would be some changes. 
The LMS would do the authentication for the user, that's correct. But I think we should tell directus which LMS service we allow. Otherwise every LMS could issue a token.

**Reply by NilsBaumgartner1994** [reply]
> I am thinking.. Actually Directus does not need to authenticate anything, since the LMS should authenticate the user and pass the info to the Directus instance. So, wouldn´t it be possible to create custom endpoints, bypassing directus auth and issuing a token to the client?

But the main problem currently would be, that we need to allow directus in an iFrame. As far as I know directus does not support being in an iFrame (#12939). Maybe @rijkvanzanten can tell us more if an iFrame is possible?
I tried it one but maybe I haven't configured the Headers correctly.

**Reply by paescuj** [reply]
> 1. On user interaction, the LMS launches an oauth2 flow that reaches Directus.
> 2. Directus would validate the issuer of that request and on success create an auth request, including a redirect url to the tool, and send that back to the LMS.

Please note that Directus is not an OAuth provider but a client (at least for now & I'm not sure whether this is planned for the future).
Therefore I think something like this would be the more sensible approach:

> I am thinking.. Actually Directus does not need to authenticate anything, since the LMS should authenticate the user and pass the info to the Directus instance. So, wouldn´t it be possible to create custom endpoints, bypassing directus auth and issuing a token to the client?

---

> But the main problem currently would be, that we need to allow directus in an iFrame

@NilsBaumgartner1994 I'm sorry, but I still do not fully understand how the integration should look like? Should the entire Directus app be displayed in the LMS via an iFrame? Would users then use that "iFramed" app to view / modify items?

**Reply by rijkvanzanten** [reply]
@NilsBaumgartner1994 If possible, would you be able to share a screen recording of another piece of software that you're already using like this? :)

**Reply by NilsBaumgartner1994** [reply]
> @NilsBaumgartner1994 If possible, would you be able to share a screen recording of another piece of software that you're already using like this? :)

Sure. In the following i will display how we could use the LTI. We have a System called Stud.IP which is our main Web-Tool for our students.

1. We then authenticate there:
<img width="1162" alt="Bildschirmfoto 2023-01-12 um 23 14 51" src="https://user-images.githubusercontent.com/33295329/212192595-9d0edf7d-31b4-434f-8517-6d6b8c4ae82a.png">

2. Students / Teacher will then head to a course. At the menu you can see the added "My-LTI-Tool"
<img width="1206" alt="Bildschirmfoto 2023-01-12 um 23 15 53" src="https://user-images.githubusercontent.com/33295329/212192753-d356f087-ceac-48df-9df8-502173d5c0f3.png">

3. By selecting this tab. You should see your LTI-Tool. In the example i have my own Frontend-App shown.
<img width="1206" alt="Bildschirmfoto 2023-01-12 um 23 17 29" src="https://user-images.githubusercontent.com/33295329/212193009-d7c57b5d-2efa-404d-bbc7-3bd3ef9a4ad5.png">

4. Unfortunatly using a directus SSO will result in an error page. This is because directus does not support an iFrame or i dont know how to configure it properly. Using the default email+password method works here but not the SSO.
<img width="1162" alt="Bildschirmfoto 2023-01-12 um 23 19 30" src="https://user-images.githubusercontent.com/33295329/212193289-3854dc85-c748-45bc-bdb9-2209f049838c.png">

5. You can edit the LTI integration and set the params and options.
<img width="730" alt="Bildschirmfoto 2023-01-12 um 23 20 19" src="https://user-images.githubusercontent.com/33295329/212193410-8806ce77-88e9-4668-acfe-849aa1a9cb66.png">

6. The Idea would be then, that the application would communicate with the LTI-Platform and send changes to the Platform which are selected in the Tool. For example. Uploading a solution or giving points to students.

**Comment by danielband** [comment]
There are two problems:
1. You need to implement [seamless SSO ](https://docs.directus.io/self-hosted/sso.html#seamless-sso). That means your app has to be on the same domain as Directus or on a subdomain. That said, I don´t know if this would work with an iframe.
2. Neither [LTI1.1](https://www.imsglobal.org/recipe-making-lti-1-tool-providers) nor LTI1.3 works the way you are using it.

EDIT: Wait, what is displayed in the iframe? Directus Admin or a another app (that is "build" by Directus)? It is a bit of an inception. From my understanding you are using Directus to the "build" apps (Tools), display those Tools in your LMS, but handle everything related to "auth flow" validation, access control etc. in Directus instead of the Tool itself.

Like I stated before, the LTI-Tool does not need to authenticate the user. The LMS has to pass the user details to it. That can be seen in Step 5 of your flow "Nutzerdaten an LTI-Tool senden" -> "Send user data to LTI-Tool". So in my opinion, this does not affect Directus authentication at all and both LTI1.1 and LTI1.3 could be implemented with custom endpoint extensions. Basically you would have to create your Directus auth token in your custom endpoint, store it on the client side and pass the token with every request to Directus.

Please somebody correct me, in case this is all nonsense.

**Reply by NilsBaumgartner1994** [reply]
Seems logical. But the main problem then would be to use Directus in an iFrame. Any idea what Env Variables has to be set for this?

**Reply by danielband** [reply]
I would say that you should not run Directus Admin in an iframe 😅 You make an app that calls your Directus instance via API.

**Reply by NilsBaumgartner1994** [reply]
So you say the idea would be:
1. Open the Frontend
2. Send the LTI-Data to custom Directus endpoints
3. Verify in Directus and send token back
4. Frontend proceeds with token as normal
?

Yes this could be a possible way. Then we would need two endpoints: LTI 1.1 and LTI 1.3 ?

**Reply by NilsBaumgartner1994** [reply]
Do the LMS sends the information which LTI it uses?

**Reply by NilsBaumgartner1994** [reply]
Does the LMS sends the information which LTI it uses?

**Reply by danielband** [reply]
> So you say the idea would be:
> 
>     1. Open the Frontend
> 
>     2. Send the LTI-Data to custom Directus endpoints
> 
>     3. Verify in Directus and send token back
> 
>     4. Frontend proceeds with token as normal
>        ?
> 
> 
> Yes this could be a possible way. Then we would need two endpoints: LTI 1.1 and LTI 1.3 ?

Yes, that´s what I am thinking of. You would have at least one endpoint for each method. LTI1.1 should be easier to implement. According to [this ressource](https://www.imsglobal.org/recipe-making-lti-1-tool-providers) the LMS should pass the LTI version as launch message.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
